cmd2.py

```diff
@@ -969,25 +970,46 @@ def _surround_ansi_escapes(prompt, start="\x01", end="\x02"):
         return result
 
     def pseudo_raw_input(self, prompt):
-        """copied from cmd's cmdloop; like raw_input, but accounts for changed stdin, stdout"""
+        """
+        began life as a copy of cmd's cmdloop; like raw_input but
+
+        - accounts for changed stdin, stdout
+        - if input is a pipe (instead of a tty), look at self.echo
+          to decide whether to print the prompt and the input
+        """
 
         # Deal with the vagaries of readline and ANSI escape codes
         safe_prompt = self._surround_ansi_escapes(prompt)
 
         if self.use_rawinput:
             try:
-                line = sm.input(safe_prompt)
+                if sys.stdin.isatty():
+                    line = sm.input(safe_prompt)
+                else:
+                    line = sm.input()
+                    if self.echo:
+                        sys.stdout.write('{}{}\n'.format(safe_prompt,line))
             except EOFError:
                 line = 'eof'
         else:
```

Comment:
This **else** case is there honestly just because it is there in ``cmd`` and because ``cmd`` defines the ``raw_input`` attribute.  It is never used explicitly in ``cmd2`` other than some unit testing "just in case".